File: code/sdk/android/cloudsheet-Android-sdk/src/main/kotlin/com/opencloudsheet/storage/workbook/create/OneDrivePersonalWorkbookCreator.kt
package com.opencloudsheet.storage.workbook.create

import android.util.Log
import com.google.gson.Gson
import com.opencloudsheet.auth.IAuthenticator
import com.opencloudsheet.model.file.ICloudFile
import com.opencloudsheet.model.file.OneDriveFile
import com.opencloudsheet.constants.OneDriveClient
import com.opencloudsheet.constants.OneDriveConstants
import com.opencloudsheet.response.onedrive.DriveItem
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject

/**
 * Creates Excel workbooks in the current user's personal OneDrive.
 *
 * Uses the Microsoft Graph API /me endpoint to create workbooks in the authenticated user's drive.
 * This differs from other implementations that may operate on different users' drives.
 *
 * @see <a href="https://learn.microsoft.com/en-us/graph/api/driveitem-post-children?view=graph-rest-1.0">Microsoft Graph: Create Drive Item</a>
 */
class OneDrivePersonalWorkbookCreator(
    private val authProvider: IAuthenticator,
    private val oneDriveClient: OneDriveClient
) : ICreateWorkBook {

    companion object {
        private const val TAG = "OneDrivePersonalWorkbookCreator"
        private val gson = Gson()
    }

    override suspend fun createWorkbook(folder: ICloudFile?, name: String): ICloudFile {
        if (name.isBlank()) {
            Log.e(TAG, "Sheet name cannot be blank")
            throw IllegalArgumentException("Sheet name cannot be blank")
        }

        if (folder != null && !folder.isFolder()) {
            Log.e(TAG, "Provided folder must be a folder, not a file: ${folder.getId()}")
            throw IllegalArgumentException("Provided folder must be a folder, not a file")
        }

        val token = authProvider.getAuthToken()
        if (token == null) {
            Log.e(TAG, "No authentication token available")
            throw IllegalStateException("No authentication token available")
        }

        val fileName = if (name.endsWith(".xlsx", ignoreCase = true)) name else "$name.xlsx"

        val url = buildCreateFileUrl(folder)

        val requestBody = buildExcelFileRequestBody(fileName)

        val request = Request.Builder()
            .url(url)
            .addHeader("Authorization", "Bearer $token")
            .addHeader("Content-Type", "application/json")
            .post(requestBody)
            .build()

        return withContext(Dispatchers.IO) {
            oneDriveClient.instance.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    Log.e(TAG, "HTTP request failed: ${response.code} - ${response.message}")
                    throw Exception("Failed to create Excel file: HTTP ${response.code} - ${response.message}")
                }

                val responseBody = response.body?.string()
                if (responseBody.isNullOrEmpty()) {
                    Log.e(TAG, "Empty response body received")
                    throw Exception("Empty response body")
                }

                parseCreatedFile(responseBody)
            }
        }
    }

    private fun buildCreateFileUrl(folder: ICloudFile?): String {
        val baseUrl = OneDriveConstants.BASE_MS_GRAPH_URL.toString()

        return if (folder == null) {
            // Create in root: POST /me/drive/root/children
            "$baseUrl/me/drive/root/children"
        } else {
            // Create in specific folder: POST /me/drive/items/{folder-id}/children
            "$baseUrl/me/drive/items/${folder.getId()}/children"
        }
    }

    private fun buildExcelFileRequestBody(fileName: String): RequestBody {
        val jsonBody = JSONObject().apply {
            put("name", fileName)
            put("file", JSONObject())
            put("@microsoft.graph.conflictBehavior", "fail")
        }

        return jsonBody.toString().toRequestBody("application/json".toMediaType())
    }

    private fun parseCreatedFile(responseBody: String): ICloudFile {
        val driveItem = gson.fromJson(responseBody, DriveItem::class.java)
        return OneDriveFile(driveItem)
    }
}

File: code/sdk/ios/OpenCloudSheet/Sources/OpenCloudSheet/Storage/Workbook/Create/OneDrivePersonalWorkbookCreator.swift
//
//  OneDrivePersonalWorkbookCreator.swift
//  OpenCloudSheet
//
//  Creates Excel workbooks in the current user's personal OneDrive
//

import Foundation

class OneDrivePersonalWorkbookCreator: ICreateWorkBook {
    private let authProvider: IAuthenticator
    private let oneDriveClient: OneDriveClient

    init(authProvider: IAuthenticator, oneDriveClient: OneDriveClient) {
        self.authProvider = authProvider
        self.oneDriveClient = oneDriveClient
    }

    func createWorkbook(folder: ICloudFile?, name: String) async throws -> ICloudFile {
        guard !name.isEmpty else {
            throw NSError(domain: "OneDrivePersonalWorkbookCreator", code: -1,
                         userInfo: [NSLocalizedDescriptionKey: "Sheet name cannot be blank"])
        }

        if let folder = folder, !folder.isFolder() {
            throw NSError(domain: "OneDrivePersonalWorkbookCreator", code: -2,
                         userInfo: [NSLocalizedDescriptionKey: "Provided folder must be a folder, not a file"])
        }

        guard let token = try await authProvider.getAuthToken() else {
            throw NSError(domain: "OneDrivePersonalWorkbookCreator", code: -3,
                         userInfo: [NSLocalizedDescriptionKey: "No authentication token available"])
        }

        let fileName = name.hasSuffix(".xlsx") ? name : "\(name).xlsx"

        let url = buildCreateFileUrl(folder: folder)
        let requestBody = try buildExcelFileRequestBody(fileName: fileName)

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.addValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = requestBody

        let (data, response) = try await oneDriveClient.session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 201 else {
            throw NSError(domain: "OneDrivePersonalWorkbookCreator", code: -4,
                         userInfo: [NSLocalizedDescriptionKey: "Failed to create Excel file"])
        }

        return try parseCreatedFile(data: data)
    }

    private func buildCreateFileUrl(folder: ICloudFile?) -> URL {
        let baseUrl = OneDriveConstants.baseMSGraphURL

        if let folder = folder {
            return URL(string: "\(baseUrl)/me/drive/items/\(folder.getId())/children")!
        } else {
            return URL(string: "\(baseUrl)/me/drive/root/children")!
        }
    }

    private func buildExcelFileRequestBody(fileName: String) throws -> Data {
        let jsonBody: [String: Any] = [
            "name": fileName,
            "file": [:],
            "@microsoft.graph.conflictBehavior": "fail"
        ]

        return try JSONSerialization.data(withJSONObject: jsonBody)
    }

    private func parseCreatedFile(data: Data) throws -> ICloudFile {
        let driveItem = try JSONDecoder().decode(DriveItem.self, from: data)
        return OneDriveFile(item: driveItem)
    }
}
